cpp/include/cudf/lists/detail/scatter.cuh

```diff
@@ -715,64 +721,82 @@ struct list_child_constructor {
 }
 
-}  // namespace
-
 /**
- * @brief Scatters lists into a copy of the target column
- * according to a scatter map.
- *
- * The scatter is performed according to the scatter iterator such that row
- * `scatter_map[i]` of the output column is replaced by the source list-row.
- * All other rows of the output column equal corresponding rows of the target table.
- *
- * If the same index appears more than once in the scatter map, the result is
- * undefined.
+ * @brief General implementation of scattering into list column
  *
- * The caller must update the null mask in the output column.
+ * Scattering `source` into `target` according to `scatter_map`.
+ * The view order of `source` and `target` can be specified by
+ * `source_vector` and `target_vector` respectively.
  *
- * @tparam SourceIterator must produce list_view objects
  * @tparam MapIterator must produce index values within the target column.
  *
+ * @param source_vector A vector of `unbound_list_view` into source column
+ * @param target_vector A vector of `unbound_list_view` into target column
+ * @param scatter_map_begin Start iterator of scatter map
+ * @param scatter_map_end End iterator of scatter map
+ * @param source Source column view
+ * @param target Target column view
  * @param stream CUDA stream used for device memory operations and kernel launches.
  * @param mr Device memory resource used to allocate the returned column's device memory
  * @return New lists column.
  */
 template <typename MapIterator>
-std::unique_ptr<column> scatter(
-  column_view const& source,
+std::unique_ptr<column> scatter_impl(
+  rmm::device_uvector<unbound_list_view> const& source_vector,
+  rmm::device_uvector<unbound_list_view>& target_vector,
   MapIterator scatter_map_begin,
   MapIterator scatter_map_end,
+  column_view const& source,
   column_view const& target,
-  rmm::cuda_stream_view stream        = 0,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
 {
-  auto const num_rows = target.size();
-
-  if (num_rows == 0) { return cudf::empty_like(target); }
-
-  auto const child_column_type = lists_column_view(target).child().type();
-
   assert_same_data_type(source, target);
 
-  using lists_column_device_view = cudf::detail::lists_column_device_view;
-  using unbound_list_view        = cudf::lists::detail::unbound_list_view;
+  auto const child_column_type = lists_column_view(target).child().type();
 
-  auto const source_device_view = column_device_view::create(source, stream);
-  auto const source_vector      = list_vector_from_column(unbound_list_view::label_type::SOURCE,
-                                                     lists_column_device_view(*source_device_view),
-                                                     stream,
-                                                     mr);
+#ifndef NDEBUG
```

Comment:
Please remove the debug code. I believe scatter is used as a primitive in other cudf algorithms. This code will be difficult to maintain and frustrating to ignore when _not_ specifically debugging this scatter. It also increases the size of the debug version of libcudf.so which is already getting to be too large for the debugger.